Week 4 Day 2 update (#69)
microsoft/win-student-devs: docs/30DaysOfPWA/platforms-practices/02.md

# 4.2: Best Practices for PWA Engagement

**Author: Zach Teutsch [@devteutsch](https://twitter.com/devteutsch)**

![Placeholder Banner Only. Replace when final assets ready.](_media/day2.png)

> Welcome to Week 4, Day 2 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 will be all about some of the best practices and technologies you can use to really polish up your PWA.

Author: ```suggestion
Welcome to Week 4, Day 2 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can turn our PWAs into quality application experiences. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.
```

### A Quick Note

> This article is about the best practices for engagement with PWAs. If you're looking for the basics on push notifications and badging in PWAs, check out the [Notifying Your Users of Upates]() article from Week 2.

Reviewer: Missing link here.

Reviewer 2: Also tiny mis-spelling here Updates instead of Upates

Author: ```suggestion
This article is about the best practices for engagement with PWAs. If you're looking for the basics on push notifications and badging in PWAs, check out the [Notifying Your Users of Updates](../advanced-capabilities/07.md) article from Week 2.
```

## Today's Plan
As we briefly touched on in [Day 1](01.md), a carefully thought out engagement strategy can be one of the cornerstones of a great PWA. With the "one codebase" nature of PWA, it's also easy to keep your engagement strategy consistent across different devices and platforms. 

Today, we're going to take a look at how PWAs stack up in terms of engagement and see how we can build on what we already learned about push notifications to get started off on the right foot with engagement.

## Comparing PWA Engagement
You may not have taken the dive into making your first PWA, but others have! Engagement with PWAs has been a success story for several companies that have experimented with redesigning their web experiences as PWAs. [PWAstats.com](https://aka.ms/learn-pwa/30days-4.2/pwastats.com) is a great resource for seeing how PWAs can improve many factors of your app, including engagement. Some cherry-picked figures related to engagement:

* **Starbucks doubled daily active users with their PWA**
* **Tinder's PWA saw engagement increase across the board**

> * **Forbes saw a 2x increase in session lenth and 20% increase in impressesions through their PWA**

Reviewer 3: ```suggestion
* **Forbes saw a 2x increase in session length and 20% increase in impressions through their PWA**
```

> You may not have taken the dive into making your first PWA, but others have! Engagement with PWAs has been a success story for several companies that have experimented with redesigning their web experiences as PWAs. [PWAstats.com](https://www.pwastats.com/) is a great resource for seeing how PWAs can improve many factors of your app, including engagement. Some cherry-picked figures related to engagement:

Reviewer 2: Should we call out some of the reasons mentioned in the resources these PWAs have had massive user engagement?

Author: Conscious on length here, so gonna leave this out and let them standalone.

Granted, progressive web apps aren't a magic, one-size-fits all solution for engagement, but they do give you all the tools you need for building a re-engageable application.

## Improving Engagement with Actionable Push Notifications
We've seen some numbers. Now let's take a look at how we can take advantage of [PWA notification capabilities](https://aka.ms/learn-pwa/30days-4.2/docs.microsoft.com/microsoft-edge/progressive-web-apps-chromium/how-to/notifications-badges) to improve engagement on our app.

### Content Comes First
Before you even think about sending push notifications, make sure your PWA has enough engageable content and reusability to even justify sending a push notification in the first place. If there's nothing to do on your app, why prompt a user to visit?

### What Does "Actionable" Mean?

> A great place to start with push notifications for your PWA is to make sure every notification has an action associated to it. When a user recieves a notification, there should be something that the user can **do** that is directly related to your push. This could mean:

Reviewer 3: ```suggestion
A great place to start with push notifications for your PWA is to make sure every notification has an action associated to it. When a user receives a notification, there should be something that the user can **do** that is directly related to your push. This could mean:
```

* Some new content to read or engage with
* A message that the user can reply to
* A new feature for the user to try out

A quality notification will give the user something to do on your PWA, not just have them open the application for the sake of opening it.

## Let's See Some Code!

We can build on the snippets that we saw in the [Notifying Your Users of Updates](../advanced-capabilities/07.md) article to keep our notifications actionable in practice. If you haven't see the notifications snippets from that post, go check them out and then come back to this.

Using the [Notification](https://developer.mozilla.org/en-US/docs/Web/API/Notification/Notification) specification, we can take advantage of the `data` property to associate our notification with any structured data that we want.

So, if we want to associate our notification with a specific location in our PWA, the notification payload we send could look something like this:

```javascript
{
    title: "Our Notification Title",
    body: "Our Notifcation Message",
    icon: "path/to/icon.png",
    data: {
        path: "<relative path for our notification>"
    }
}
```

> Now, we can adjust the push listener to make sure the displayed Notification captures our new `data` property:

Author: ```suggestion
Now, we can adjust the push listener in our service worker to make sure the displayed Notification captures our new `data` property:
```

```javascript
self.addEventListener('push', (event) => {
    const notificationData = JSON.parse(event.data.text());

    event.waitUntil(
        self.registration.showNotification(notificationData.title, {
            body: notificationData.message,
            icon: notificationData.icon,
            data: notificationData.data // This is the only change from week 2
        });
    );
});

> ```

Reviewer 2: Just a heads up that PWA Studio has snippets for some of these things https://github.com/pwa-builder/pwa-studio/wiki/Code-Snippets . Not sure if we would want to call that out here, but yeah

Author: Like this idea, added it after the actions snippet.
